mne/tests/test_source_tfr.py

```diff
@@ -0,0 +1,201 @@
+
+    # Changing .data re-computes .times
+    stfr.data = np.random.rand(100, 20, 5)
+    assert_array_almost_equal(
+        stfr.times, [1., 2., 3., 4., 5.])
+
+    # .data must match the number of vertices
+    pytest.raises(ValueError, attempt_assignment, stfr, 'data', [[1]])
+    pytest.raises(ValueError, attempt_assignment, stfr, 'data', None)
+
+    # .data much match number of dimensions
+    pytest.raises(ValueError, attempt_assignment, stfr, 'data', np.arange(100))
+    pytest.raises(ValueError, attempt_assignment, stfr_kernel, 'data',
+                  [np.arange(100)])
+    pytest.raises(ValueError, attempt_assignment, stfr_kernel, 'data',
+                  [[[np.arange(100)]]])
+
+    # .shape attribute must also work when ._data is None
+    stfr._kernel = np.zeros((2, 2))
+    stfr._sens_data = np.zeros((2, 5, 3))
+    stfr._data = None
+    assert_equal(stfr.shape, (2, 5, 3))
+
+    # bad size of data
+    stfr = _fake_stfr()
+    data = stfr.data[:, :, np.newaxis, :]
+    with pytest.raises(ValueError, match='3 dimensions for SourceTFR'):
+        SourceTFR(data, stfr.vertices)
+    stfr = SourceTFR(data[:, :, 0, 0], stfr.vertices, 0, 1)
+    assert stfr.data.shape == (data.shape[0], data.shape[1], 1)
+
+
+@requires_h5py
+def test_io_stfr_h5():
+    """Test IO for stfr files using HDF5."""
+    for stfr in [_fake_stfr(), _fake_kernel_stfr()]:
+        tempdir = _TempDir()
+        pytest.raises(ValueError, stfr.save, op.join(tempdir, 'tmp'),
+                      ftype='foo')
+        out_name = op.join(tempdir, 'tempfile')
+        stfr.save(out_name, ftype='h5')
+        stfr.save(out_name, ftype='h5')  # test overwrite
+        # TODO: no read_source_tfr yet
+
+
+def test_stfr_resample():
+    """Test resampling SourceTFR."""
+    stfr_ = _fake_stfr()
+    kernel_stfr_ = _fake_kernel_stfr()
+
+    for stfr in [stfr_, kernel_stfr_]:
+        stfr_new = deepcopy(stfr)
+        o_sfreq = 1.0 / stfr.tstep
+        # note that using no padding for this stfr reduces edge ringing...
+        stfr_new.resample(2 * o_sfreq, npad=0)
+        assert (stfr_new.data.shape[-1] == 2 * stfr.data.shape[-1])
+        assert (stfr_new.tstep == stfr.tstep / 2)
+        stfr_new.resample(o_sfreq, npad=0)
+        assert (stfr_new.data.shape[-1] == stfr.data.shape[-1])
+        assert (stfr_new.tstep == stfr.tstep)
+        assert_array_almost_equal(stfr_new.data, stfr.data, 5)
+
+
+def test_stfr_crop():
+    """Test cropping of SourceTFR."""
+    stfr = _fake_stfr()
+    kernel_stfr = _fake_kernel_stfr()
+
+    for inst in [stfr, kernel_stfr]:
+        copy_1 = inst.copy()
+        assert_allclose(copy_1.crop(tmax=0.8).data, inst.data[:, :, :9])
+        # FIXME: the next line doesn't work for kernel stfr nor kernel stc
+        # assert_allclose(copy_1.times, inst.times[:9])
+
+        copy_2 = inst.copy()
+        assert_allclose(copy_2.crop(tmin=0.2).data, inst.data[:, :, 2:])
+        assert_allclose(copy_2.times, inst.times[2:])
```

Comment:
the reference you compare with is the later `assert actual == expected`. Same thing `assert_allclose(actual, expected)